Serialize dNR rule identifiers in content extension rule lists

elijahsawyers opened the pull request:
#### 2c0f6fd37a286e28225a59ea83c9e6fb15a43298
<pre>
Serialize dNR rule identifiers in content extension rule lists
<a href="https://bugs.webkit.org/show_bug.cgi?id=297155">https://bugs.webkit.org/show_bug.cgi?id=297155</a>
<a href="https://rdar.apple.com/157890918">rdar://157890918</a>

Reviewed by Brian Weinstein.

This patch makes it possible to specify an identifier for a content extension rule. This is working
toward implementing the onRuleMatchedDebug API for dNR, which isn&apos;t currently possible because the
rule identifiers are destroyed when the dNR rules are converted into content extension rules.

The idea behind this patch is that we store the rule identifier with its serialized action, meaning
that we cannot combine actions, which does have slight memory implications. However, most dNR rules
contain conditions and wouldn&apos;t have their actions combined during compilation anyways. As for the
identifiers themselves, we have a 100k rule limit, and each identifier is a 32 bit unsigned integer,
so at most, we&apos;re using 0.4MB more memory.

The serialized actions now have the form:

[
  one byte for the variant,
  X bytes for extra information (like string size and selector for css-display-none actions),
  four bytes for the identifier,
  ...
]

There&apos;s a lot more to do to get this to work, like implementing the dNR API and wiring up this new
notification. Also, there are smaller things such as conditional serializing of the identifiers and
adding more information to the object that&apos;s included with the published notification.

See sub-tasks of: <a href="https://rdar.apple.com/71867958">rdar://71867958</a>

Wrote new tests to validate that the rules are being serialized as expected. Tests for the
notification will be written once we wire up the functionality to the web extensions controller.
See: <a href="https://rdar.apple.com/157878200">rdar://157878200</a>.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::findFirstIgnoreRule const):
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::serializeActions):
(WebCore::ContentExtensions::compileRuleList):
* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRule):
(WebCore::ContentExtensions::loadEncodedRules):
* Source/WebCore/contentextensions/ContentExtensionRule.cpp:
(WebCore::ContentExtensions::ContentExtensionRule::ContentExtensionRule):
(WebCore::ContentExtensions::DeserializedAction::deserialize):
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::ContentExtensionRule::identifier const):
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy const):
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::contentRuleListMatchedRule):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentRuleListMatchedRule):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, RuleIdentifiers)):

Canonical link: <a href="https://commits.webkit.org/298570@main">https://commits.webkit.org/298570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90b3915bf505892e91d14e9dec931be860f29d5f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35488 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121882 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66476 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fd921333-fe47-426c-bc90-9db557cd00df) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44075 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87989 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42610 "Exiting early after 60 failures. 20791 tests run. 60 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9707d063-72d1-4649-b27f-39233c9322ef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68394 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27996 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22062 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65553 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107943 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98252 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125035 "Built successfully") | | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114362 "Found 1 new JSC stress test failure: stress/catch-from-function-inlining-simd-function.js.ftl-eager-no-cjit (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32055 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96745 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100135 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96531 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41790 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19667 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38643 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18526 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42610 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48199 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/142692 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42077 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36884 "Found 7354 new JSC stress test failures: airjs-tests.yaml/stress-test.js.lockdown, basic-tests.yaml/stress-test.js.lockdown, cdjs-tests.yaml/main.js.lockdown, cdjs-tests.yaml/motion_test.js.lockdown, cdjs-tests.yaml/red_black_tree_test.js.lockdown, cdjs-tests.yaml/reduce_collision_set_test.js.lockdown, microbenchmarks/Float32Array-matrix-mult.js.lockdown, microbenchmarks/Int16Array-bubble-sort-with-byteLength.js.lockdown, microbenchmarks/Int16Array-bubble-sort.js.lockdown, microbenchmarks/Int16Array-load-int-mul.js.lockdown ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45411 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43784 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1014,6 +1014,11 @@
+// When this is enabled, CurrentContentRuleListFileVersion and currentDeclarativeNetRequestRuleTranslatorVersion need to be bumped.
+#if !defined(ENABLE_DNR_ON_RULE_MATCHED_DEBUG)
+#define ENABLE_DNR_ON_RULE_MATCHED_DEBUG 0
+#endif
```
I'm not sure what is the right answer, but which do you think is better - this flow, or having a new valve for `CurrentContentRuleListFileVersion`/`currentDeclarativeNetRequestRuleTranslatorVersion` when the feature flag is on so that it gets automatically bumped when we turn the feature flag on?

elijahsawyers replied:
I didn't want to have a new value with the flag turned on in case we bump the version[s] before this gets turned on. If that happened, the value[s] would be outdated.

b-weinstein commented on Source/WebCore/contentextensions/ContentExtensionParser.cpp:
```diff
@@ -313,6 +325,9 @@ static Expected<Vector<ContentExtensionRule>, std::error_code> loadEncodedRules(
     Vector<ContentExtensionRule> ruleList;
 
+#if ENABLE(DNR_ON_RULE_MATCHED_DEBUG)
+    size_t rulesWithIdentifiers = 0;
```
Nit: From the variable name, it seems like this would be a set of rules. `numberOfRulesWithIdentifiers`/`rulesWithIdentifiersCount` might make this a bit more clear.

b-weinstein commented on Source/WebCore/contentextensions/ContentExtensionRule.cpp:
```diff
@@ -87,7 +94,17 @@ DeserializedAction DeserializedAction::deserialize(std::span<const uint8_t> seri
+
+    uint32_t identifier = location;
+#if ENABLE(DNR_ON_RULE_MATCHED_DEBUG)
+    // FIXME: <rdar://157879177> We shouldn't unconditionally deserialize an identifier here, as all rule list do not serialize identifiers.
```
Nit: `as all rule lists do not`....

b-weinstein commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -758,6 +762,16 @@ static bool isUnsupportedWebExtensionNavigation(API::NavigationAction& navigatio
+    if (!m_navigationState)
+        return;
+
+    // FIXME: rdar://157878200 (dNR: send a message from the navigation client to the web extensions controllers when a content extension rule matches)
```
How similar do you think this will be to `NavigationState::NavigationClient::contentRuleListNotification`? If we didn't have a feature flag, do you think it would make more sense to pass an optional `WebCore::ContentRuleListMatchedRule` to that method?

elijahsawyers replied:
They serve different purposes. `contentRuleListNotification` is called once per request, whereas we call this new notification for each rule that matches for a request.

I think the separation is fine, unless there's a strong reason to combine them.

b-weinstein commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
```diff
@@ -64,6 +64,12 @@
+            argument = "--gtest_filter=ContentExtensionTest.RuleIdentifiers"
+            isEnabled = "YES">
+         </CommandLineArgument>
+      </CommandLineArguments>
```
This part of the change should be reverted.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
```diff
@@ -3424,6 +3433,31 @@ TEST_F(ContentExtensionTest, IgnoreFollowingRules)
+    ASSERT_EQ(data.actions.size(), 17u);
+
+    ASSERT_EQ(data.actions[0], 0u);
+    ASSERT_EQ(reinterpretCastSpanStartTo<const uint32_t>(data.actions.subvector(1, 4).span()), 1u);
```
I think this could use a comment on how this gets to the identifier (and what each of these checks are testing),

elijahsawyers committed:
Serialize dNR rule identifiers in content extension rule lists
https://bugs.webkit.org/show_bug.cgi?id=297155
rdar://157890918

Reviewed by Brian Weinstein.

This patch makes it possible to specify an identifier for a content extension rule. This is working
toward implementing the onRuleMatchedDebug API for dNR, which isn't currently possible because the
rule identifiers are destroyed when the dNR rules are converted into content extension rules.

The idea behind this patch is that we store the rule identifier with its serialized action, meaning
that we cannot combine actions, which does have slight memory implications. However, most dNR rules
contain conditions and wouldn't have their actions combined during compilation anyways. As for the
identifiers themselves, we have a 100k rule limit, and each identifier is a 32 bit unsigned integer,
so at most, we're using 0.4MB more memory.

The serialized actions now have the form:

[
  one byte for the variant,
  X bytes for extra information (like string size and selector for css-display-none actions),
  four bytes for the identifier,
  ...
]

There's a lot more to do to get this to work, like implementing the dNR API and wiring up this new
notification. Also, there are smaller things such as conditional serializing of the identifiers and
adding more information to the object that's included with the published notification.

See sub-tasks of: rdar://71867958

Wrote new tests to validate that the rules are being serialized as expected. Tests for the
notification will be written once we wire up the functionality to the web extensions controller.
See: rdar://157878200.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::findFirstIgnoreRule const):
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::serializeActions):
(WebCore::ContentExtensions::compileRuleList):
* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRule):
(WebCore::ContentExtensions::loadEncodedRules):
* Source/WebCore/contentextensions/ContentExtensionRule.cpp:
(WebCore::ContentExtensions::ContentExtensionRule::ContentExtensionRule):
(WebCore::ContentExtensions::DeserializedAction::deserialize):
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::ContentExtensionRule::identifier const):
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy const):
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::contentRuleListMatchedRule):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentRuleListMatchedRule):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, RuleIdentifiers)):

Canonical link: https://commits.webkit.org/298570@main
